overlord/snapstate/handlers_prereq_test.go

```diff
@@ -732,3 +732,93 @@ func (s *prereqSuite) TestDoPrereqSnapdNoRevision(c *C) {
 	c.Check(chg.Err(), ErrorMatches, `cannot perform the following tasks:\n.*- test \(cannot install system snap "snapd": no snap revision available as specified\)`)
 }
+
+func (s *prereqSuite) TestPreReqContentAttrsNotSatisfied(c *C) {
+	snapstate.AutoAliases = func(*state.State, *snap.Info) (map[string]string, error) {
+		return nil, nil
+	}
+	s.AddCleanup(func() { snapstate.AutoAliases = nil })
+
+	st := s.state
+	st.Lock()
+
+	// mock the snap which is the default-provider
+	mockInstalledSnap(c, st, `name: some-snap`, false)
+
+	t := s.state.NewTask("prerequisites", "test")
+	t.Set("snap-setup", &snapstate.SnapSetup{
+		SideInfo: &snap.SideInfo{
+			RealName: "foo",
+			Revision: snap.R(33),
+		},
+		Channel:            "beta",
+		Base:               "core",
+		PrereqContentAttrs: map[string][]string{"some-snap": {"this-does-not-match"}},
+	})
+	chg := s.state.NewChange("sample", "...")
+	chg.AddTask(t)
+	st.Unlock()
+
+	s.se.Ensure()
+	s.se.Wait()
+
+	st.Lock()
+	defer st.Unlock()
+
+	// As we are not seeding, expect no messages being logged for the task. Instead
+	// the update should have added a lot of tasks, so the resulting change should
+	// contain the update for some-snap.
+	c.Check(chg.Err(), IsNil)
+	c.Check(len(chg.Tasks()) > 1, Equals, true)
+
+	// Expect the initial prerequisites task to have completed
+	c.Check(chg.Tasks()[0].Kind(), Equals, "prerequisites")
+	c.Check(chg.Tasks()[0].Status(), Equals, state.DoneStatus)
+	c.Check(chg.Tasks()[0].Log(), HasLen, 0)
+}
+
+func (s *prereqSuite) TestPreReqContentAttrsNotSatisfiedSeeding(c *C) {
+	snapstate.AutoAliases = func(*state.State, *snap.Info) (map[string]string, error) {
+		return nil, nil
+	}
+	s.AddCleanup(func() { snapstate.AutoAliases = nil })
+
+	// mock that we are not seeded
+	r := snapstatetest.MockDeviceModelAndMode(DefaultModel(), "install")
+	defer r()
+
+	st := s.state
+	st.Lock()
+	st.Set("seeded", false)
+
+	// mock the snap which is the default-provider
+	mockInstalledSnap(c, st, `name: some-snap`, false)
+
+	t := s.state.NewTask("prerequisites", "test")
+	t.Set("snap-setup", &snapstate.SnapSetup{
+		SideInfo: &snap.SideInfo{
+			RealName: "foo",
+			Revision: snap.R(33),
+		},
+		Channel:            "beta",
+		Base:               "core",
+		PrereqContentAttrs: map[string][]string{"some-snap": {"this-does-not-match"}},
+	})
+	chg := s.state.NewChange("sample", "...")
+	chg.AddTask(t)
+	st.Unlock()
+
+	s.se.Ensure()
+	s.se.Wait()
+
+	st.Lock()
+	defer st.Unlock()
+
+	// We expect the change to complete without error, even if updating a snap is not
+	// allowed during seeding. Instead what we expect, is a task log message noting that
+	// updating the snap was not allowed due to the seeding stage.
+	c.Check(chg.Err(), IsNil)
+	c.Assert(chg.Tasks(), HasLen, 1)
+	c.Assert(chg.Tasks()[0].Log(), HasLen, 1)
+	c.Check(chg.Tasks()[0].Log()[0], testutil.Contains, `too early for operation, device not yet seeded or device model not acknowledged`)
```

Comment:
I guess here we could also look for the message "cannnot update %q during se" etc?

Reply:
Fixed!